controller/k8s/watcher_test.go

```diff
@@ -51,4 +58,34 @@ func TestWatcher(t *testing.T) {
 			t.Fatalf("Expected error, got nil")
 		}
 	})
+	t.Run("Returns error if the watcher does not initialize in the time limit", func(t *testing.T) {
+		resource := &resourceToWatchImpl{}
+		var lock sync.RWMutex
+		didRetry := false
```

Comment:
The `didRetry` var is going to be set to true the first time `testWatchInit` is called, which happens immediately. So this var being `true` isn't an indication that retries happened. I think you should instead switch to using an int to count the number of function calls, and assert that it has been called the correct number of times before being timed out.

That will also allow you to use an atomic int, which means you don't need to mess around with the mutex inside this test. So I'd rewrite this as:

```
		watchInitCalls := int32(0)
		expectedCalls := int32(3)
		testWatchInit := func(stop <-chan struct{}) error {
			atomic.AddInt32(&watchInitCalls, 1)
			return errors.New("mock error")
		}
```